nifi-external/nifi-kafka-connect/nifi-kafka-connector/src/main/java/org/apache/nifi/kafka/connect/StatelessNiFiSourceTask.java

```diff
@@ -0,0 +1,298 @@
+
+        // Kafka Connect currently gives us no way to determine the number of partitions that a given topic has.
+        // Therefore, we have no way to partition based on an attribute or anything like that, unless we left it up to
+        // the dataflow developer to know how many partitions exist a priori and explicitly set an attribute in the range of 0..max,
+        // but that is not a great solution. Kafka does support using a Simple Message Transform to change the partition of a given
+        // record, so that may be the best solution.
+        final Integer topicPartition = null;
+
+        final String topic;
+        if (topicNameAttribute == null) {
+            topic = topicName;
+        } else {
+            final String attributeValue = flowFile.getAttribute(topicNameAttribute);
+            topic = attributeValue == null ? topicName : attributeValue;
+        }
+
+        final ConnectHeaders headers = new ConnectHeaders();
+        if (headerAttributeNamePattern != null) {
+            // TODO: When we download/create the dataflow, create a hash of it. Then save that state. When we do it next time,
+            //       compare the hash to the last one. If changed, need to trigger connect framework to tell it that the config has changed.
+            //       Would be done via Source/Sink Context.
+            //       Or perhaps we should include the flow JSON itself in the configuration... would require that we string-ify the JSON though. This would be the cleanest, though. Would be optional.
+            //       We can just document that you either include it inline, or you don't make changes to the dataflow; instead, save as a separate dataflow and update task to point to the new one.
+
+            for (final Map.Entry<String, String> entry : flowFile.getAttributes().entrySet()) {
+                if (headerAttributeNamePattern.matcher(entry.getKey()).matches()) {
+                    final String headerName = entry.getKey();
+                    final String headerValue = entry.getValue();
+                    headers.add(headerName, headerValue, Schema.STRING_SCHEMA);
+                }
+            }
+        }
+
+        final Object key = keyAttributeName == null ? null : flowFile.getAttribute(keyAttributeName);
+        final Schema keySchema = key == null ? null : Schema.STRING_SCHEMA;
+        final Long timestamp = System.currentTimeMillis();
+
+        return new SourceRecord(partitionMap, componentState, topic, topicPartition, keySchema, key, valueSchema, contents, timestamp, headers);
+    }
+
+    @Override
+    public void commitRecord(final SourceRecord record, final RecordMetadata metadata) throws InterruptedException {
+        super.commitRecord(record, metadata);
+
+        final long unacked = unacknowledgedRecords.decrementAndGet();
+        logger.debug("SourceRecord {} committed; number of unacknowledged FlowFiles is now {}", record, unacked);
+
+        if (unacked < 1) {
```

Comment:
Since this is usually invoked from another thread, there can be an edge case where:

- the unacknowledgedRecords is already decremented, making it 0, but the result is still not acknowledged
- poll is invoked at this point on the task thread, and we trigger the dataflow while the previous result is still not acknowledged.

Reply:
I think this scenario is okay. When that happens, the call to `dataflow.trigger()` will happen. This will return a `DataflowTriggerResult` immediately, but the call to `DataflowTriggerResult.getResult()` will block until the dataflow completes in the background. But the dataflow won't actually run in the background until the previous one is acknowledged. So the second triggering of the dataflow doesn't actually happen until the previous one has been acknowledged.

Unless i'm overlooking something on the connect side that makes this problematic?